### PR TITLE
Two fixes to weapon states

### PR DESCRIPTION
**Pull request by Baezon**
When the weapon state is changed for the purposes of creating particle sources, while switching to the new one it checks if the old one was `WeaponState::INVALID` and if so attempts to create a `WeaponState::NORMAL` particle source. Given that the normal, not-invalid case creates a new source for the new state directly afterwards anyway, this is redundant and causes NORMAL weapon states (that is, particles spawned by primaries) to have double the particle sources.

Additionally an invalid `wp->homing_object` is `&obj_used_list` and not `nullptr`, this was causing all dumbfired missiles to not create any particle sources at all.

### CODE REVIEW

File: code/weapon/weapons.cpp [hunk 1]
```diff
@@ -4929,7 +4911,7 @@ static void weapon_update_state(weapon* wp)
 			infree_flight = true;
 		}
 		else if (lifetime >= fl2f(wip->free_flight_time) &&
-			(lifetime - Frametime) <= fl2f(wip->free_flight_time) && wp->homing_object != nullptr)
+			(lifetime - Frametime) <= fl2f(wip->free_flight_time) && wp->homing_object != &obj_used_list)
```
Comment (JohnAFernandez) [inline code/weapon/weapons.cpp:4914]: We'll need a comment about the &obj_used_list to make sure it's clear to others in the future.  I myself would feel very confused if I saw that.

**Review by JohnAFernandez — CHANGES_REQUESTED** [review]

File: code/weapon/weapons.cpp (hunk 1 above)
Comment (Baezon) [reply code/weapon/weapons.cpp:4914]: It's still weird to me, but its something about linked lists. It is used all over the code though, so it seems strange to single out these spots, should I be commenting all of them?

Comment (JohnAFernandez) [reply code/weapon/weapons.cpp:4914]: *looks around* Just here, I think.

**Review by JohnAFernandez — APPROVED** [review]

**Comment by JohnAFernandez** [comment]
Sorry I left this languishing for a little bit.

**Pull request merged**